docs/config.md
# Configuration file for the webserver

This file is used to configure the webserver. It is used to define the server's behavior and settings. 

For our project we wanted to use a configuration file that we knew already existed and was used by many people. We chose to use the Nginx configuration file. However the Nginx doesnt support the cgi scripts. So we did a hybrid of Nginx and Apache configuration files.  

See here for an overview of different configuration files: [config_files.md](docs/config_files.md).  
At first we thought that Caddy would be a good candidate for a simple config file, but there is a lot under the hood, that we would not implement in our project. You can see here for an overview of the Caddy configuration file: [caddy_config.md](docs/caddy_configuration_docs.md).  

So in the end we decided that nginx would be our best bet. We used the nginx configuration file as a base and added the cgi script support from the apache configuration file.

An example of our configuration file with all cases covered would be:
```nginx
http {
    keepalive_timeout 5;
    
    server {
        listen 4244;
        listen 4245;
        server_name myWebserver someWebserver;
        root www;
        client_max_body_size 100000000;
        
        # Error pages
        error_page {
            400 www/error_pages/400.html
            403 www/error_pages/403.html
            404 www/error_pages/404.html
            405 www/error_pages/405.html
            418 www/error_pages/418.html
            500 www/error_pages/500.html
            502 www/error_pages/502.html
        }

        # Location blocks
        location /42 {
            return 301 http://42berlin.de/;
        }

        location /43 {
            autoindex on;
        }

        location /go {
            return 301 /here/index.html;
        }

        location /uploads {
            file_upload on;
        }

        # CGI Configuration
        cgi {
            cgi_path_alias /cgi-bin "/cgi-bin"
            cgi_extension .pl /bin/perl
            cgi_extension .py /bin/python3
            upload_dir www/uploads
        }

        # Accepted Methods
        servers GET POST DELETE PUT {
            deny all;
        }
    }

    # Second server
    server {
        listen 4246;
        server_name myWebserver someWebserver;
        root ./www/html;
    }

    # Third server
    server {
        listen 4247;
        server_name myWebserver someWebserver;
		root ./www/html;
    }

}
```

This would populate a base struct with the following values:
```cpp
struct BaseConf {
  std::size_t maxConnections;
  int requestTimeout;   // read timeout - this is used!
  int responseTimeout;  // write timeout - this is used!

  int keepalive_timeout;      // The maximum time to keep an idle connection open.
  size_t maxBodySize;      // defaults to 0 for infinite if not specified
  std::map<std::string, std::string> headers;
  bool autoindex; // will present a list of files in a directory
  bool file_server; // it will serve everything it is default!
  std::vector<std::string> acceptedMethods;
  std::map<int, std::string> error_pages; // Map of error codes to custom error pages 
  std::string upload_dir; // Directory to store uploaded files

  CGIData cgiData;	
};
```
and a struct dor each directive block:
```cpp
struct Directive {
  std::vector<std::string> acceptedMethods; // List of accepted HTTP methods GET, POST, PUT, DELETE, etc.
//   std::string redirect_to;
  bool autoindex; // will present a list of files in a directory
  bool file_upload; // will allow file uploads
  std::string upload_dir; // Directory to store uploaded files
  std::string alias; // for rerouting
  bool internal; // for internal only pages
  std::pair<int, std::string> return_directive; // Pair to store return directive
  std::map<int, std::string> error_pages; // Map of error codes to custom error pages location level 
};
```
The cgi has its own block:
```cpp
struct CGIData {
  std::pair<std::string, std::string> cgi_path_alias; // Pair to store CGI path alias
  std::map<std::string, std::string> cgi_extensions; // Map of file extensions to CGI interpreters paths
  std::string upload_dir;
};
```

since the cgi uses scripts, I will have on the top of each script a shebang with the path to the interpreter. Therefore it is not necessary to store the interpreter in the configuration file.

docs/config_files.md
# CONFIGURATION FILES

the below is from the copilot ai, not tested. just for an idea of the syntax of different web servers configuration files.

## Nginx Configuration example

```nginx
http {
    server {
        listen 80;
        server_name example.com;

        # Default error pages
        error_page 404 /404.html;
        error_page 500 502 503 504 /50x.html;

        # Limit client body size
        client_max_body_size 1M;

        location / {
            root /var/www/html;
            index index.html index.htm;
            autoindex off; # Turn off directory listing
        }

        location /images/ {
            root /var/www/html;
            autoindex on; # Turn on directory listing
        }

        location /redirect {
            return 301 http://example.com/;
        }

        location /upload {
            acceptedMethods POST {
                deny all;
            }
        }
    }

    server {
        listen 8080;
        server_name another.example.com;

        location / {
            root /var/www/another;
            index index.html index.htm;
        }
    }
}
```

## Caddy Configuration example

```caddyfile
example.com {
    # Choose the port and host
    bind 0.0.0.0
    port 80

    # Default error pages
    handle_errors {
        @404 {
            expression {http.error.status_code} == 404
        }
        rewrite @404 /404.html
        file_server
    }

    # Limit client body size
    request_body {
        max_size 1MB
    }

    # Setup routes
    route / {
        root * /var/www/html
        file_server browse
    }

    route /images/* {
        root * /var/www/html
        file_server browse
    }

    route /redirect {
        redir http://example.com/
    }

    route /upload {
        method POST
        respond "Upload endpoint"
    }
}

another.example.com {
    bind 0.0.0.0
    port 8080

    route / {
        root * /var/www/another
        file_server
    }
}
```
### example caddy path with CGI
```bash
http://localhost:2019 {
    root * /home/lbrusa/Documents/caddyHTTP
    file_server

    @add path /add
    cgi @add /path/to/cgi-bin/add
}
```

## Apache Configuration example

```apache
<VirtualHost *:80>
    ServerName example.com

    # Default error pages
    ErrorDocument 404 /404.html
    ErrorDocument 500 /50x.html

    # Limit client body size
    LimitRequestBody 1048576

    # Setup routes
    DocumentRoot /var/www/html
    <Directory /var/www/html>
        Options -Indexes # Turn off directory listing
        AllowOverride None
        Require all granted
    </Directory>

    <Directory /var/www/html/images>
        Options +Indexes # Turn on directory listing
    </Directory>

    Redirect /redirect http://example.com/

    <Location /upload>
        <LimitExcept POST>
            Require all denied
        </
<Location /upload>
        <LimitExcept POST>
            Require all denied
        </LimitExcept>
    </Location>
</VirtualHost>

<VirtualHost *:8080>
    ServerName another.example.com

    DocumentRoot /var/www/another
    <Directory /var/www/another>
        Options -Indexes
        AllowOverride None
        Require all granted
    </Directory>
</VirtualHost>
```

- Nginx: Uses server blocks to define servers, location blocks for routes, and directives like client_max_body_size to limit client body size.
- Caddy: Uses a simple, human-readable configuration file with route blocks for routes and request_body for limiting client body size.
- Apache: Uses <VirtualHost> blocks to define servers, <Directory> and <Location> blocks for routes, and directives like LimitRequestBody to limit client body size


src/Config.cpp
#include "Config.hpp"
#include "HTTPServer.hpp"
#include "ServerData.hpp"
#include "Constants.hpp"
#include "debug.h"
#include <cstdlib>
#include <fstream>
#include <sstream>
#include <stdexcept>
#include <string>

/** Initialize the static variables */
std::vector<ServerData> Config::servers;
Config *Config::instance_ = NULL;
std::string Config::_filename = "config/default.conf"; // Default filename
std::map<uint16_t, ServerData *> Config::port_map_;

/**
 * @brief Constructor for the Config class
 *
 * It will get a filename and read the configuration file
 * or default to a default configuration file if none is provided.
 */

Config::Config(std::string filename) {
	if (filename.empty()) {
		_filename = Constants::default_config_file;
	} else {
		_filename = filename;
	}

  Parser::parse(_filename, servers, port_map_);
  if (!validate()) {
	cleanup();
	debuglog(RED, "Configuration validation failed");
	throw std::runtime_error("Invalid configuration");
  }
  
  debuglog(GREEN, "\n\nConfig initialized with %zu servers\n\n",
           servers.size());
}

Config::Config(const Config &) {}
Config &Config::operator=(const Config &) { return *this; }
Config::~Config() {}

void Config::cleanup() {
  delete instance_;
  instance_ = NULL;
  debuglog(YELLOW, "Server config data destroyed");
}

void Config::initialize(std::string &config_file) {
  if (instance_ == NULL) {
    instance_ = new Config(config_file);
  } 
}

const std::vector<ServerData> &Config::getServerData() {
	if (_filename.empty()) {
	  _filename = Constants::default_config_file;
	}
	if (instance_ == NULL) {
	  instance_ = new Config(Config::_filename);
	 	}
	return Config::servers;
  }

const std::vector<ServerData> &Config::getServerData(char *config_file) {
  if (config_file == NULL) {
	  _filename = Constants::default_config_file;
  } 
  if (instance_ == NULL) {
    instance_ = new Config(Config::_filename);
	
  }
  return Config::servers;
}

const ServerData *Config::getConfigByPort(uint16_t port) {
  if (instance_ == NULL) {
    instance_ = new Config(Config::_filename);
   
  }
  std::map<uint16_t, ServerData *>::const_iterator it = port_map_.find(port);
  return (it != port_map_.end()) ? it->second : NULL;
}

bool Config::validate() {
	if (servers.size() == 0) {
		debuglog(RED, "Configuration error: Empty server array");
		return false;
	}
	for (size_t i = 0; i < servers.size(); ++i) {
		if (servers[i].ports.empty()) {
		debuglog(RED, "Configuration error: No ports specified");
		return false;
		} else if (servers[i].server_names.empty()) {
		debuglog(RED, "Configuration error: No server names specified");
		return false;
		} else if (servers[i].root.empty()) {
		debuglog(RED, "Configuration error: Empty root directory");
		return false;
		} 
	}
	return true;
}


/*
Config::Config(std::string filename) {
  (void)filename;

  // First server configuration
  ServerData server1;
  server1.keepalive_timeout = 5;
  server1.autoindex = true;
  Location location;
  location.return_directive = std::make_pair(301, "http://42berlin.de/");
  server1.location_blocks["/42"] = location;
  location = Location();
  location.autoindex = true;
  server1.location_blocks["/43"] = location;
  location = Location();
  location.return_directive = std::make_pair(301, "/here/index.html");
  server1.location_blocks["/go"] = location;
  location = Location();
  location.file_upload = true;
  server1.location_blocks["/uploads"] = location;
  server1.ports.push_back(SERVER_PORT);
 // server1.ports.push_back(4245);
  /* test the servernames with curl -H "Host: myWebserver"
   * http://localhost:4244/ or curl -H "Host: someWebserver"
   * http://localhost:4244/ or curl -H "Host: myWebserver"
   * http://localhost:4245/ or curl -H "Host: someWebserver"
   * http://localhost:4245/ or nc localhost 4244 and GET / HTTP/1.1 Host:
   * myWebserver
   */
/*
server1.server_names.push_back("myWebserver");
server1.server_names.push_back("someWebserver");
server1.root = "html/www1";
server1.index = "index.html";
server1.error_pages.insert(
    std::make_pair(400, server1.root + "/error_pages/400.html"));
server1.error_pages.insert(
    std::make_pair(403, server1.root + "/error_pages/403.html"));
server1.error_pages.insert(
    std::make_pair(404, server1.root + "/error_pages/404.html"));
server1.error_pages.insert(
    std::make_pair(405, server1.root + "/error_pages/405.html"));
server1.error_pages.insert(
    std::make_pair(418, server1.root + "/error_pages/418.html"));
server1.error_pages.insert(
    std::make_pair(500, server1.root + "/error_pages/500.html"));
server1.error_pages.insert(
    std::make_pair(502, server1.root + "/error_pages/502.html"));

server1.upload_dir = "http/www1/uploads";
server1.maxBodySize = 100000000;
server1.acceptedMethods.push_back("GET");
server1.acceptedMethods.push_back("POST");
server1.acceptedMethods.push_back("DELETE");
server1.acceptedMethods.push_back("PUT");
server1.cgiData.cgi_path_alias = std::make_pair("/cgi-bin", "/cgi-bin");
server1.cgiData.upload_dir = "www/uploads";
// server1.cgiData = data;
// Second server configuration
ServerData server2;
server2.ports.push_back(4246);
server2.server_names.push_back("myWebserver");
server2.server_names.push_back("someWebserver");
server2.root = "http/www2/";

// Third server configuration
ServerData server3;
server3.ports.push_back(4247);
server3.server_names.push_back("myWebserver");
server3.server_names.push_back("someWebserver");
server3.root = "http/www3/";

// Add all servers to servers
servers.push_back(server1);
servers.push_back(server2);
servers.push_back(server3);
// map ports to server configurations
for (size_t i = 0; i < servers.size(); ++i) {
  for (size_t j = 0; j < servers[i].ports.size(); ++j) {
    port_map_[servers[i].ports[j]] = &servers[i];
  }
}
debuglog(YELLOW, "Config initialized with %zu servers", servers.size());
}*/
/**
 * @brief Prints all server configurations in the servers vector
 *
 * This function prints detailed information about all server configurations
 * including ports, server names, locations, CGI settings, etc.
//
*/
